docker/utils/utils.py

```diff
@@ -88,13 +88,17 @@ def build_file_list(root):
     return files
 
 
-def create_archive(root, files=None, fileobj=None, gzip=False):
+def create_archive(root, files=None, fileobj=None, gzip=False,
+                   extra_files=None):
     if not fileobj:
         fileobj = tempfile.NamedTemporaryFile()
     t = tarfile.open(mode='w:gz' if gzip else 'w', fileobj=fileobj)
     if files is None:
         files = build_file_list(root)
     for path in files:
+        if path in [e[0] for e in extra_files]:
```

Comment:
Nitpicking, but `any(e[0] == path for e in extra_files)` won't build the whole list in memory. Maybe even building a `set(e[0] for e in extra_files)` before to avoid iterating too many times.